Sources/Sapi.php

```diff
@@ -0,0 +1,318 @@
+					$is_os = PHP_OS_FAMILY === 'Darwin';
+					break;
+				// This may result in false positives because 'linux' is very broad.
+				case self::OS_LINUX:
+					$is_os = PHP_OS_FAMILY === 'Linux';
+					break;
+			}
+
+			if ($is_os) {
+				return true;
+			}
+		}
+
+		// Last ditch effort.
+		return PHP_OS_FAMILY === $os;
+	}
+
+	/**
+	 * Checks If we are runnig a CGI instance.
+	 * 
+	 * @return bool True if we are running under CGI, false otherwise.
+	 */
+	public static function isCGI(): bool
+	{
+		return isset($_SERVER['SERVER_SOFTWARE']) && strpos(php_sapi_name(), 'cgi') !== false;
+	}
+
+	/**
+	 * Checks If we are runnig a CLI (shell/cron) instance.
+	 * 
+	 * @return bool True if we are running under CLI, false otherwise.
+	 */
+	public static function isCLI(): bool
+	{
+		return empty($_SERVER['REQUEST_METHOD']);
+	}
+
+	/**
+	 * Checks if the server is able to support case folding.
+	 * 
+	 * @see https://www.w3.org/TR/charmod-norm/#definitionCaseFolding
+	 * @return bool True if it does, false otherwise
+	 */
+	public static function isCaseFolding(): bool
+	{
+		return ord(strtolower(chr(138))) === 154;
+	}
+
+	/**
+	 * A bug in some versions of IIS under CGI (older ones) makes cookie setting not work with Location: headers.
+	 * 
+	 * @return bool True if it does, false otherwise
+	 */
+	public static function needsLoginFix(): bool
+	{
+		return self::isCGI() && self::isSoftware(self::SERVER_IIS);
+	}
+
+	/**
+	 * (Re)initializes some $context values that need to be set dynamically.
+	 */
+	public static function load(): void
+	{
+		// This determines the server... not used in many places, except for login fixing.
+		Utils::$context['server'] = [
+			'is_iis' => self::isSoftware(self::SERVER_IIS),
+			'is_apache' => self::isSoftware(self::SERVER_APACHE),
+			'is_litespeed' => self::isSoftware(self::SERVER_LITESPEED),
+			'is_lighttpd' => self::isSoftware(self::SERVER_LIGHTTPD),
+			'is_nginx' => self::isSoftware(self::SERVER_NGINX),
+			'is_cgi' => self::isCGI(),
+			'is_windows' => self::isOS(self::OS_WINDOWS),
+			'is_mac' => self::isOS(self::OS_MAC),
+			'iso_case_folding' => self::isCaseFolding(),
+			'needs_login_fix' => self::needsLoginFix()
+		];
+	}
+
+	/**
+	 * Locates the most appropriate temporary directory.
+	 *
+	 * Systems using `open_basedir` restrictions may receive errors with
+	 * `sys_get_temp_dir()` due to misconfigurations on servers. Other
+	 * cases sys_temp_dir may not be set to a safe value. Additionally
+	 * `sys_get_temp_dir` may use a readonly directory. This attempts to
+	 * find a working temp directory that is accessible under the
+	 * restrictions and is writable to the web service account.
+	 *
+	 * Directories checked against `open_basedir`:
+	 *
+	 * - `sys_get_temp_dir()`
+	 * - `upload_tmp_dir`
+	 * - `session.save_path`
+	 * - `cachedir`
+	 *
+	 * @return string Path to a temporary directory.
+	 */
+	public static function getTempDir(): string
+	{
+		return Config::getTempDir();
```

Comment:
I tried.  But we call the getTempDir very early and it breaks Config from loading (autoloader isn't running yet).  I wanted to take another stab at it someday, but the most logical place is for it to be in the Sapi class since it is talking to the server about where to locate a temporary working directory.

If you want it in the Config class only we can. I had wrote all the changes when I came across this, so I simply just pulled it back and updated the one reference in Config class that was broken.

Reply:
No, its fine this way. What you did makes sense. It's not quite ideal in terms of the purpose of the Sapi class, but it is the best approach in this case.